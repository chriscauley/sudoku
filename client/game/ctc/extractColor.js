const color_map = {
  '#A3E048': 'green',
  '#F7D038': 'yellow',
  '#34BBE6': 'blue',
  '#CFCFCF': 'border',
  '#E6261F': 'red',
  '#D23BE7': 'magenta',
  '#EB7532': 'orange',
  '#000000': 'text',
  '#FFFFFF': 'bg',
}

export default (color) => {
  color = color.toUpperCase()
  if (!color_map[color]) {
    console.warn('missing color', color)
    return 'border'
  }
  return color_map[color]
}
