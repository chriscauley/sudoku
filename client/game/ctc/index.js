import classNames from 'classnames'
import { vector, dxy2text } from '../Geo'
export { default as buildArrows } from './buildArrows'

const css = {
  xy: (xy) => `x-${xy[0]} y-${xy[1]}`,
}

const extractColor = (color) => {
  const color_map = {
    '#A3E048': 'green',
    '#F7D038': 'yellow',
    '#34BBE6': 'blue',
    '#CFCFCF': 'border',
    '#E6261F': 'red',
    '#D23BE7': 'magenta',
    '#EB7532': 'orange',
    '#000000': 'text',
    '#FFFFFF': 'white',
  }

  color = color.toUpperCase()
  if (!color_map[color]) {
    console.warn('missing color', color)
    return 'gray'
  }
  return color_map[color]
}

// currently only thermometers & diagonals
// forked thermometer #H7n7NhH26M
// lots of theremometers #J7pMMgrLL3
export const buildLines = (lines, board) => {
  const thermometer_colors = {}
  const forced_ends = {} // used in "shall we play a game" #Z4gGYPtBWNw

  const { W, H } = board.geo
  const processedLines = lines
    .filter((l) => l.wayPoints.length)
    .map((line) => {
      const { wayPoints } = line
      const color = extractColor(line.color)
      const cells = wayPoints.map((wp) => {
        const xy = wp.map((i) => Math.floor(i)).reverse()
        const index = board.geo.xy2index(xy)
        if (wp[0] % 0.5 || wp[1] % 0.5) {
          forced_ends[index] = color
        }
        return { xy, index }
      })
      if (cells.length === 2 && cells[0].index + cells[1].index === 90) {
        const diff = Math.abs(cells[0].index - cells[1].index)
        let xys = [],
          direction
        if (diff === 90) {
          xys = vector.connect([0, 0], [W - 1, H - 1])
          direction = 'up'
        } else if (diff === 72) {
          xys = vector.connect([0, H - 1], [W - 1, 0])
          direction = 'down'
        } else {
          // not actually a diagonal, just two cells that sum to 90
          return { cells, type: 'thermo', color }
        }
        return {
          type: 'diagonal',
          direction,
          cells: xys.map((xy) => ({
            xy,
            index: board.geo.xy2index(xy),
            className: classNames('line-diagonal', direction, 'color-' + color),
          })),
        }
      }
      return { cells, type: 'thermo', color }
    })

  // fill in the cells between the old "waypoint" cells, which are just the corners
  board.extras.lines = processedLines.map((line) => {
    let last_xy
    const line_xys = []
    if (line.type === 'thermo') {
      line.cells.forEach((point) => {
        if (last_xy) {
          vector
            .connect(last_xy, point.xy)
            .slice(1)
            .map((xy) => line_xys.push(xy))
        } else {
          line_xys.push(point.xy)
        }
        last_xy = point.xy
      })
      line.cells = line_xys.map((xy) => ({
        index: board.geo.xy2index(xy),
        className: 'line-thermo ' + 'color-' + line.color,
        xy,
      }))
    }

    return line
  })

  // TODO replace lines with canvas, which will make this calss obsolete
  board.extras.lines.forEach((line) => {
    const color = line.color
    line.cells.forEach((cell) => {
      if (forced_ends[cell.index] === color) {
        cell.className += ' forced_end'
      }
    })
  })

  board.extras.lines.forEach((line) => {
    const { color } = line
    thermometer_colors[color] = thermometer_colors[color] || {}
    if (line.type === 'thermo') {
      let last_point
      line.cells.forEach((point) => {
        const { index } = point
        thermometer_colors[color][index] =
          thermometer_colors[color][index] || []
        if (last_point) {
          point.className +=
            ' from from-' +
            dxy2text[vector.sign(vector.subtract(last_point.xy, point.xy))]
          last_point.className +=
            ' to to-' +
            dxy2text[vector.sign(vector.subtract(point.xy, last_point.xy))]
          thermometer_colors[color][index].push(last_point.index)
          thermometer_colors[color][last_point.index].push(index)
        }
        last_point = point
      })
    }
    line.cells.forEach((point) => (point._className = point.className))
  })

  buildThermometers(board, thermometer_colors, forced_ends)
}

const fixThermometer = (thermometer, u) => {
  // right now this only handles single forked bulbs because that's the only exception I've encountered
  const index2 = thermometer[u.index][0]
  delete thermometer[u.index]
  thermometer[index2].push(u.index)
}

const buildThermometers = (board, _colors, forced_ends) => {
  // build out the "this cell is greater than these" map used in checking thermometers
  const thermometers = {}
  board.extras.thermometers = []
  board.extras.marks.forEach((u) => {
    if (u.type !== 'thermo') {
      return
    }
    thermometers[u.color] = thermometers[u.color] || []
    const matched = thermometers[u.color].find((t) => t[u.index])
    if (matched) {
      fixThermometer(matched, u)
      return
    }
    let depth = 0
    const _thermo = {}
    const thermometer = _colors[u.color]
    if (!(thermometer && thermometer[u.index])) {
      console.warn('Cannot find thermometer')
      return
    }
    const chase = (index, last) => {
      depth++
      let is_end = forced_ends[index] === u.color
      !is_end &&
        thermometer[index].forEach((index2) => {
          if (index2 === last || depth > 100) {
            return
          }
          _thermo[index2] = _thermo[index2] || []
          _thermo[index2].push(index)
          is_end = true
          chase(index2, index)
        })
    }
    chase(u.index, undefined, [])
    thermometers[u.color].push(_thermo)
    board.extras.thermometers.push(_thermo)
  })
}

// example puzzles:
// H7n7NhH26M - crazy thermometers
// 7Qh3tBm4mj - ceiling and floor
export const buildMarks = (marks, geo) => {
  marks = marks.map((mark) => {
    const center = mark.center.reverse()
    const ratio = mark.width / mark.height
    const xy = center.map((n) => Math.floor(n))
    const index = geo.xy2index(xy)
    let type = mark.rounded ? 'thermo' : 'square'
    let orientation = ''
    if (ratio < 1) {
      type = 'wall'
      orientation = 'h-split'
    } else if (ratio > 1) {
      type = 'wall'
      orientation = 'v-split'
    }
    const color = extractColor(mark.backgroundColor)
    const className = classNames(
      `mark ${orientation} ${type} color-${color}`,
      css.xy(xy),
      { rounded: mark.rounded },
    )
    return {
      index,
      offset: [xy[0] - center[0], xy[1] - center[1]],
      text: mark.text,
      xy,
      type,
      orientation,
      color,
      className,
      _className: className,
    }
  })

  const mark_indexes = {}
  marks.forEach((u) => (mark_indexes[u.index] = u))
  marks.forEach((mark) => {
    if (!geo.xyInGrid(mark.xy)) {
      mark.type = 'gutter'
      return
    }
    mark.next_to = geo
      .index2pawn(mark.index)
      .map((i) => mark_indexes[i])
      .filter(Boolean)
    mark.is_end = mark.next_to.length === 1
  })
  return marks
}
