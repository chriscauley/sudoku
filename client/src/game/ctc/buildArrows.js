import { cloneDeep } from 'lodash'

import { vector } from '../Geo'
import css from '@unrest/css'

export const fromCTC = ({ wayPoints }) => {
  wayPoints = cloneDeep(wayPoints)
  // waypoints use yx coordinates, so flip them
  const wp1 = wayPoints[wayPoints.length - 2].reverse()
  const wp2 = wayPoints[wayPoints.length - 1].reverse()
  const average = vector.add(wp1, wp2).map((n) => n / 2)
  const d = vector.subtract(wp1, wp2).map((n) => Math.abs(n))

  const out = {
    xy: average.map((n) => Math.floor(n)),
    dxy: [-Math.sign(wp1[0] - wp2[0]), -Math.sign(wp1[1] - wp2[1])],
  }
  const long = (out.long = !!d.find((n) => n > 1))
  if (long) {
    // spans two cells, average will not work. Use head waypoint as xy and mark as long
    out.xy = wp2.map((n) => Math.floor(n))
    out.long = true
  }
  out.className = [out, 'arrow']
  return out
}

export default (board) => {
  const { arrows = [] } = board.options.ctc
  const { geo } = board
  board.extras.arrows = []
  arrows.forEach((ctc_arrow) => {
    if (ctc_arrow.wayPoints.length < 2) {
      return
    }
    const arrow = fromCTC(ctc_arrow)
    arrow.index = geo.xy2index(arrow.xy)
    if (arrow.long) {
      board.options.ctc.lines.push(ctc_arrow)
    }
    if (geo.xyInGrid(arrow.xy)) {
      board.extras.arrows.push(arrow)
    } else {
      console.warn('TODO: gutter arrow at', arrow.xy)
    }
  })
}
