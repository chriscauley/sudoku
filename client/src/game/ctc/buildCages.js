import { cloneDeep } from 'lodash'

export default (board) => {
  board.extras.cages = cloneDeep(board.options.ctc.cages).map((cage) => {
    cage.indexes = []
    cage.first = { index: Infinity }
    cage.last = { index: 0 }
    cage.cells = cage.cells.map((xy) => {
      xy = xy.reverse()
      const index = board.geo.xy2index(xy)
      const cell = { xy, index, className: 'cage' }
      cage.indexes.push(index)
      if (cage.first.index > cell.index) {
        cage.first = cell
      }
      if (cage.last.index < cell.index) {
        cage.last = cell
      }
      return cell
    })
    cage.first.className += ' cage-first'
    cage.last.className += ' cage-last'
    cage.first.text = cage.last.text = cage.value
    cage.cells.forEach((cell) => {
      Object.entries(board.geo._text2dindex).forEach(([text, dindex]) => {
        if (!cage.indexes.includes(cell.index + dindex)) {
          cell.className += ' cage-' + text
        }
        cell._className = cell.className
      })
    })
    return cage
  })
}
