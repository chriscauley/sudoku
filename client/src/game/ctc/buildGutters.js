import { range } from 'lodash'

class Gutter {
  constructor(board, g) {
    this.name = board.geo.g2text(g)
    this.is_column = g % 2 === 0
    this.SIZE = board.geo[this.is_column ? 'H' : 'W']
    this.values = range(this.SIZE).map(() => undefined)
    this.class = 'gutter gutter-' + this.name
  }
  get = (index) => this.values[index]
  set = (index, value) => (this.values[index] = value)
}

// this may need to be in ctc.js
export default (marks, board) => {
  board.gutters = range(4).map((g) => new Gutter(board, g))
  marks.forEach((mark) => {
    const xy = mark.xy
    if (board.geo.xyInGrid(xy)) {
      console.warn('Mark found in grid. Marks currently only work for gutter.')
      return
    }
    const gi = board.geo.xy2gi(xy)
    if (!gi) {
      console.warn('no gutter matching', xy)
      return
    }
    board.gutters[gi[0]].set(gi[1], mark.text)
  })
}
