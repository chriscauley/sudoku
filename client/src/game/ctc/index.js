export { default as buildArrows } from './buildArrows'
export { default as buildCages } from './buildCages'
export { default as buildGutters } from './buildGutters'
export { default as buildLines } from './buildLines'
export { default as buildMarks } from './buildMarks'
