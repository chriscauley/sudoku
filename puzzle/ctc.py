import datetime
from server.utils import curl
from bs4 import BeautifulSoup

import re
import requests
import json

_exid = 'cracking-the-cryptic.web.app/sudoku/'
match_external_id = re.compile(r'.*cracking-the-cryptic.web.app/sudoku/(\w+)')

CTC_URL = "https://firebasestorage.googleapis.com/v0/b/sudoku-sandbox.appspot.com/o/"
def _get_ctc(url):
    request = requests.get(url)
    request.raise_for_status()
    token = request.json()['downloadTokens']
    return requests.get(f"{url}?alt=media&token={token}").text

def fetch_ctc(slug):
    data = json.loads(curl(CTC_URL + slug, getter=_get_ctc, name=slug, force=True))
    for row in data.get('cells', []):
        for cell in row:
            if not cell.get('pencilMarks', None):
                cell.pop('pencilMarks', None)
            if not cell.get('candidates', None):
                cell.pop('candidates', None)
    return data

def update_video(video, Puzzle):
    values = {}
    video_url = f"https://www.youtube.com/watch?v={video.external_id}"
    html = curl(video_url)
    soup = BeautifulSoup(html, features="html.parser")
    title = soup.find("meta", {'property': 'og:title'})
    title = title and title['content']

    description = html.split('"shortDescription":"')[1].split('",')[0]
    description = json.loads(f'"{description}"').replace('\\n','\n').replace('\"','"')
    publish_date = html.split('publishDate":"')[-1].split(r'"')[0]

    puzzle_id = video.puzzle_id
    if _exid in description and not puzzle_id:
        puzzle_external_id = description.split(_exid)[1].split(' ')[0].split('\n')[0].split('\\')[0]
        puzzle = Puzzle.objects.filter(external_id=puzzle_external_id).first()
        if not puzzle:
            puzzle = Puzzle(external_id=puzzle_external_id)
            puzzle.publish_date = video.publish_date
            puzzle.save()
            print('Puzzle created for ', video.title, repr(puzzle.external_id))
        puzzle_id = puzzle.id
    publish_date = datetime.datetime.strptime(publish_date, "%Y-%m-%d").date()
    values = dict(
        description=description,
        title=title,
        publish_date=publish_date,
        puzzle_id=puzzle_id,
    )
    for key, value in values.items():
        old = getattr(video, key)
        if value != old:
            if key == 'description':
                print('updating description', len(video.description or ''), len(description))
            else:
                print("updating", key, getattr(video, key), value)
            setattr(video, key, value)

def update_ctc(puzzle):
    if puzzle.external_id and not puzzle.data.get('ctc'):
        try:
            puzzle.data['ctc'] = fetch_ctc(puzzle.external_id)
        except:
            if puzzle.flag == 'new':
                puzzle.flag = "bad_id"

def refresh_ctc(Video, Puzzle):
    response = requests.get('https://www.youtube.com/feeds/videos.xml?channel_id=UCC-UOdK8-mIjxBQm_ot1T-Q')
    text = response.text
    soup = BeautifulSoup(text, features="html.parser")
    for entry in soup.find_all('entry'):
        links = entry.find_all('link')
        if len(links) != 1:
            print("SKIPPING: Only expected one link but got:", links)
            continue
        video_id = links[0]['href'].split('v=')[-1]
        if Video.objects.filter(external_id=video_id):
            continue
        video = Video(external_id=video_id)
        update_video(video, Puzzle)
        video.save()
        print("New video:",video_id)

