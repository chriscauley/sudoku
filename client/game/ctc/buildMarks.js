import extractColor from './extractColor'
import _className from '../className'

// example puzzles:
// H7n7NhH26M - crazy thermometers
// 7Qh3tBm4mj - ceiling and floor
export default (marks, geo) => {
  marks = marks.map((mark) => {
    const center = mark.center.reverse()
    const ratio = mark.width / mark.height
    const xy = center.map((n) => Math.floor(n))
    const index = geo.xy2index(xy)
    let type = mark.rounded ? 'thermo' : 'square'
    let orientation = ''
    if (ratio < 1) {
      type = 'wall'
      orientation = 'h-split'
    } else if (ratio > 1) {
      type = 'wall'
      orientation = 'v-split'
    }
    const color = extractColor(mark.backgroundColor)
    const className = _className(
      { xy, rounded: mark.rounded },
      `mark ${orientation} ${type} color-${color}`,
    )
    return {
      index,
      offset: [xy[0] - center[0], xy[1] - center[1]],
      text: mark.text,
      xy,
      type,
      orientation,
      color,
      className,
      _className: className,
    }
  })

  const mark_indexes = {}
  marks.forEach((u) => (mark_indexes[u.index] = u))
  marks.forEach((mark) => {
    if (!geo.xyInGrid(mark.xy)) {
      mark.type = 'gutter'
      return
    }
    mark.next_to = geo
      .index2pawn(mark.index)
      .map((i) => mark_indexes[i])
      .filter(Boolean)
    mark.is_end = mark.next_to.length === 1
  })
  return marks
}
