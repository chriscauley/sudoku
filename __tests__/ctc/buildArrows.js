import { fromCTC } from '../../client/game/ctc/buildArrows'

// ignore negative zeros, they are a quirk of jest
const cases = [
  // ctc#8D346R4gjB - diagonal gutter arrows
  [[-0.5, 4.5], [0, 4]],
  [[9.5, 2.5], [9, 3]],

  // qjGQffjL6B - these caused problems because they were in the wrong cell
  [[2.8, 7],[2.8, 6]],
  [[0.8,0],[0.8,1]],
  [[4, 4.8],[3, 4.8]],
  [[3,3.8],[4,3.8]],

  // RnT7bFHJ6h - these arrows are longer than two cells
  [[7.5,8.2],[7.5,6.5]],
  [[8.2,5.5],[6.5,5.5]],
]

test('wayPointsToArrows', () => {
  cases.forEach((wayPoints) => {
    expect(fromCTC({wayPoints}).className).toMatchSnapshot()
  })
})