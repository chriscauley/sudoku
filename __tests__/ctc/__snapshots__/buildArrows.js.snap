// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wayPointsToArrows 1`] = `"x-4 y--1 dx--1 dy-1 arrow"`;

exports[`wayPointsToArrows 2`] = `"x-2 y-9 dx-1 dy--1 arrow"`;

exports[`wayPointsToArrows 3`] = `"x-6 y-2 dx--1 dy-0 arrow"`;

exports[`wayPointsToArrows 4`] = `"x-0 y-0 dx-1 dy-0 arrow"`;

exports[`wayPointsToArrows 5`] = `"x-4 y-3 dx-0 dy--1 arrow"`;

exports[`wayPointsToArrows 6`] = `"x-3 y-3 dx-0 dy-1 arrow"`;

exports[`wayPointsToArrows 7`] = `"x-6 y-7 dx--1 dy-0 long arrow"`;

exports[`wayPointsToArrows 8`] = `"x-5 y-6 dx-0 dy--1 long arrow"`;
